numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -1086,6 +1086,95 @@ TIMEDELTA_setitem(PyObject *op, void *ov, void *vap)
 
 /* Assumes contiguous, and aligned, from and to */
 
+/* special case TIMEDELTA NPY_DATETIME_NAT handling */
+
+/**begin repeat
+ *
+ * #TYPE = TIMEDELTA#
+ * #type = npy_timedelta#
+ */
+
+static void
+@TYPE@_to_HALF(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const @type@ *ip = input;
+    npy_half *op = output;
+
+    while (n--) {
+        if (*ip == NPY_DATETIME_NAT) {
+            *op = NPY_HALF_NAN;
+        }
+        else {
+            *op = npy_float_to_half((float)(*ip));
+        }
+        ip++;
+        op++;
+    }
+}
+
+static void
+HALF_to_@TYPE@(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const npy_half *ip = input;
+    @type@ *op = output;
+
+    while (n--) {
+        *op++ = (@type@)npy_half_to_float(*ip++);
+    }
+}
+
+/**end repeat**/
+#if NPY_SIZEOF_SHORT == 2
+#define HALF_to_HALF SHORT_to_SHORT
+#elif NPY_SIZEOF_INT == 2
+#define HALF_to_HALF INT_to_INT
+#endif
+
+/**begin repeat
+ *
+ * #TOTYPE = BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
+ *           LONGLONG, ULONGLONG, FLOAT, DOUBLE, LONGDOUBLE, DATETIME,
+ *           TIMEDELTA#
+ * #totype = npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+ *           npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+ *           npy_float, npy_double, npy_longdouble,
+ *           npy_datetime, npy_timedelta#
+ */
+
+/**begin repeat1
+ *
+ * #FROMTYPE = TIMEDELTA#
+ * #fromtype = npy_timedelta#
+ */
+static void
+@FROMTYPE@_to_@TOTYPE@(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const @fromtype@ *ip = input;
+    @totype@ *op = output;
+
+    while (n--) {
+        if (*ip == NPY_DATETIME_NAT) {
+            if (*op == (npy_double)(*op)) {
+                *op = (double)NPY_NAN;
+            }
+            else if (*op == (npy_float)(*op)) {
+                *op = (float)NPY_NAN;
+            }
```

Comment:
What are you trying to do with these ifs? You shouldn't ever read `op`, only write to it.

Reply:
A form of type checking; I think I had issues using the template @type@ style things in comparison operators for some reason.

Reply:
Is there a nice / clean way to inspect types on inner loops in our template language? I guess something involving PyArray_DESCR or similar might be cooked up, although the "ugly" way seem to work.

Reply:
We could do with a better way of attaching traits to types at loop-expansion time, for sure.

One approach we've used elsewhere is just to introduce another parallel variable containing 1/0 for each branch.

Another approach I'd like to investigate, but we have no precedent for, is to use something like:

```
#define float64_is_float64
```
And then in the loop use

```C++
# ifdef @type@_is_float64
...
#endif
```

Reply:
The highest performing approach is probably just to hard code the loops by type so that no check is needed at all, as alluded to above. Just means more code.

Reply:
For now - can't you just write a separate loop for the non-floating point cases? I don't think that you need to distinguish between the different types of NaN anyway - C will cast for you just fine, although perhaps with a warning that makes travis unhappy.

Reply:
Yeah, I think that's the debate I was having with myself in the comments above.

Reply:
Right - but you were looking at it as a performance optimization, whereas in fact its required for correctness.